modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/container/response/filter/EntityExtensionContainerResponseFilter.java

```diff
@@ -0,0 +1,117 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.vulcan.internal.jaxrs.container.response.filter;
+
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.model.Company;
+import com.liferay.portal.vulcan.internal.extension.EntityExtensionHandler;
+import com.liferay.portal.vulcan.internal.extension.EntityExtensionThreadLocal;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+import java.util.Map;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ContainerResponseFilter;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.ext.ContextResolver;
+import javax.ws.rs.ext.Provider;
+import javax.ws.rs.ext.Providers;
+
+/**
+ * @author Carlos Correa
+ */
+@Provider
+public class EntityExtensionContainerResponseFilter
+	implements ContainerResponseFilter {
+
+	@Override
+	public void filter(
+			ContainerRequestContext containerRequestContext,
+			ContainerResponseContext containerResponseContext)
+		throws IOException {
+
+		Map<String, Serializable> extendedProperties =
+			EntityExtensionThreadLocal.getExtendedProperties();
+
+		if (extendedProperties == null) {
+			return;
+		}
+
+		MediaType mediaType = containerResponseContext.getMediaType();
+
+		ContextResolver<EntityExtensionHandler> contextResolver =
+			_providers.getContextResolver(
+				EntityExtensionHandler.class, mediaType);
+
+		if (contextResolver == null) {
+			return;
+		}
+
+		EntityExtensionHandler entityExtensionHandler =
+			_getEntityExtensionHandler(
+				containerResponseContext.getEntityClass(), contextResolver,
+				mediaType);
+
+		if (entityExtensionHandler == null) {
+			return;
+		}
+
+		try {
+			entityExtensionHandler.setExtendedProperties(
+				_company.getCompanyId(), containerResponseContext.getEntity(),
+				extendedProperties);
+		}
+		catch (Exception exception) {
+			_log.error(exception);
+
+			throw new WebApplicationException(exception);
+		}
+	}
+
+	private EntityExtensionHandler _getEntityExtensionHandler(
+		Class<?> clazz, ContextResolver<EntityExtensionHandler> contextResolver,
+		MediaType mediaType) {
+
+		if (clazz == null) {
+			return null;
+		}
+
+		EntityExtensionHandler entityExtensionHandler =
+			contextResolver.getContext(clazz);
+
+		if (entityExtensionHandler == null) {
+			entityExtensionHandler = _getEntityExtensionHandler(
+				clazz.getSuperclass(), contextResolver, mediaType);
```

Comment:
Did you find any case where this is needed?

Reply:
Yes, for example, for the case I am testing, the UserAccount DTO class, the value of `containerResponseContext.getEntityClass()` is `com.liferay.headless.admin.user.internal.dto.v1_0.converter.UserResourceDTOConverter$1` as the way the UserAccount is created is considered by Java as an "Inner class" of the DTO Converter `UserResourceDTOConverter`

![image](https://user-images.githubusercontent.com/32699538/180967527-d5d83ae1-f406-471e-af91-17a47a204280.png)